docs/src/tutorials/hybrid-applications-with-webpack/template-deep-dive/src-lib-rs.md

```diff
@@ -0,0 +1,306 @@
+the module `prelude` inside the crate `wasm_bindgen`) can be
+referred to without prefixing it with
+`wasm_bindgen::prelude`.
+
+For example, `#[wasm_bindgen]` could also be written as
+`#[wasm_bindgen::prelude::wasm_bindgen]`, although this is
+not recommended.
+
+One other point of interest is how we import the `cfg_if!`
+macro.
+
+```rust
+#[macro_use]
+extern crate cfg_if;
+```
+
+The `#[macro_use]` attribute imports the `cfg_if!` macro the
+same way a `use` statement imports functions.
+
+
+## 3. `wee_alloc` optional dependency
+
+```rust
+cfg_if! {
+    // When the `wee_alloc` feature is enabled, use `wee_alloc` as the global
+    // allocator.
+    if #[cfg(feature = "wee_alloc")] {
+        extern crate wee_alloc;
+        #[global_allocator]
+        static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+    }
+}
+```
+
+This code block is intended to initialize `wee_alloc` as the
+global memory allocator, but only if the `wee_alloc` feature
+is enabled in `Cargo.toml`.
+
+We immediately notice that `cfg_if!` is a macro because it
+ends in `!`, similarly to other Rust macros such as
+`println!` and `vec!`. A macro is directly replaced by other
+code during compile time.
+
+During compile time, `cfg_if!` evaluates the `if` statement.
+This tests whether the feature `wee_alloc` is present in the
+`[features]` section of `Cargo.toml` (among other possible
+ways to set it).
+
+As we saw earlier, the `default` vector in `[features]` only
+contains `"console_error_panic_hook"` and not `"wee_alloc"`.
+So, in this case, the `cfg_if!` block will be replaced by no
+code at all, and hence the default memory allocator will be
+used instead of `wee_alloc`.
+
+```rust
+extern crate wee_alloc;
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+```
+
+However, suppose `"wee_alloc"` is appended to the `default`
+vector in `Cargo.toml`. Then, the `cfg_if!` block is instead
+replaced with the contents of the `if` block, shown above.
+
+This code sets the `wee_alloc` allocator to be used as the
+global memory allocator.
+
+### What is `wee_alloc`?
+
+Reducing the size of compiled WebAssembly code is important,
+since it is often transmitted over the Internet or placed on
+embedded devices.
+
+> `wee_alloc` is a tiny allocator designed for WebAssembly
+> that has a (pre-compression) code-size footprint of only a
+> single kilobyte.
+
+[An analysis](http://fitzgeraldnick.com/2018/02/09/wee-alloc.html)
+suggests that over half of the bare minimum WebAssembly
+memory footprint is required by Rust's default memory
+allocator. Yet, WebAssembly code often does not require a
+sophisticated allocator, since it often just requests a
+couple of large initial allocations.
+
+`wee_alloc` trades off size for speed. Although it has a
+tiny code-size footprint, it is relatively slow if
+additional allocations are needed.
+
+For more details, see the
+[`wee_alloc` repository](https://github.com/rustwasm/wee_alloc).
```

Comment:
Perhaps this could also link to https://rustwasm.github.io/docs/book/reference/code-size.html as one of the recommended ways to minimize code size?